Source/Core/Resource.js

```diff
@@ -0,0 +1,534 @@
+     *
+     * @param {Resource|String} resource A Resource or a String to use when creating a new Resource.
+     * @param {Object{ options If resource is a String, these are the options passed to the Resource constructor. It is ignored otherwise.
+     *
+     * @returns {Resource} If resource is a String, a Resource constructed with the url and options. Otherwise the resource parameter is returned.
+     *
+     * @private
+     */
+    Resource.createIfNeeded = function(resource, options) {
+        if (!defined(resource)) {
+            return;
+        }
+
+        if (typeof resource === 'string') {
+            var args = defined(options) ? clone(options) : {};
+            args.url = resource;
+            resource = new Resource(args);
+        }
+
+        return resource;
+    };
+
+    defineProperties(Resource.prototype, {
+        /**
+         * Query parameters appended to the url.
+         *
+         * @memberof Resource.prototype
+         * @type {Object}
+         *
+         * @readonly
+         */
+        queryParameters: {
+            get: function() {
+                return this._queryParameters;
+            }
+        },
+
+        /**
+         * The key/value pairs used to replace template parameters in the url.
+         *
+         * @memberof Resource.prototype
+         * @type {Object}
+         *
+         * @readonly
+         */
+        templateValues: {
+            get: function() {
+                return this._templateValues;
+            }
+        },
+
+        /**
+         * The url to the resource.
+         *
+         * @memberof Resource.prototype
+         * @type {String}
+         */
+        url: {
+            get: function() {
+                return this.getUrlComponent(true, true);
+            },
+            set: function(value) {
+                var uri = new Uri(value);
+
+                if (defined(uri.query)) {
+                    var query = parseQuery(uri.query);
+                    this._queryParameters = combine(this._queryParameters, query);
+                    uri.query = undefined;
+                }
+
+                // Remove the fragment as it's not sent with a request
+                uri.fragment = undefined;
+
+                this._url = uri.toString();
+
+                if (this._isDirectory) {
+                    this._url = appendForwardSlash(this._url);
+                }
+            }
+        },
+
+        /**
+         * The file extension of the resource.
+         *
+         * @memberof Resource.prototype
+         * @type {String}
+         *
+         * @readonly
+         */
+        extension: {
+            get: function() {
+                return getExtensionFromUri(this._url);
+            }
+        },
+
+        /**
+         * True if the Resource refers to a directory. A trailing forward slash will be maintained.
+         *
+         * @memberof Resource.prototype
+         * @type {Boolean}
+         */
+        isDirectory: {
```

Comment:
I mentioned this elsewhere, but I think we can remove this as a property and just have the called pass it into `getDerivedResource`, since the caller should know better regarding how the derived resource is created.